code/go/0chain.net/smartcontract/minersc/fees.go

```diff
@@ -298,40 +287,52 @@ func (msc *MinerSmartContract) payFees(t *transaction.Transaction,
 		return "", err
 	}
 
-	minerRewards, sharderRewards, err := gn.splitByShareRatio(blockReward)
+	minerRewards, sharderRewards, err = gn.splitByShareRatio(blockReward)
 	if err != nil {
 		return "", fmt.Errorf("error splitting rewards by ratio: %v", err)
 	}
-	minerFees, sharderFees, err := gn.splitByShareRatio(fees)
+	minerFees, sharderFees, err = gn.splitByShareRatio(fees)
 	if err != nil {
 		return "", fmt.Errorf("error splitting fees by ratio: %v", err)
 	}
-	// pay random N miners
-	if err := mn.StakePool.DistributeRewardsRandN(
-		minerRewards,
-		mn.ID,
-		spenum.Miner,
-		b.GetRoundRandomSeed(),
-		gn.NumMinerDelegatesRewarded,
-		spenum.BlockRewardMiner,
-		balances,
-	); err != nil {
-		return "", err
-	}
 
-	if err := mn.StakePool.DistributeRewardsRandN(
-		minerFees,
-		mn.ID,
-		spenum.Miner,
-		b.GetRoundRandomSeed(),
-		gn.NumMinerDelegatesRewarded,
-		spenum.FeeRewardMiner,
-		balances,
-	); err != nil {
-		return "", err
+	var mn *MinerNode
+	if mn, err = getRewardedMiner(b, balances); err != nil {
+		return "", common.NewErrorf("pay_fees", "cannot get miner to reward, %v", err)
+	}
+	if mn == nil {
+		logging.Logger.Info("pay_fees, could not find miner to reward", zap.Int64("round", b.Round))
```

Comment:
should we return here?

Reply:
Why should no miner available for rewards, maybe because they have all been killed, prevent sharders from receiving rewards? And visa versa.

IMHO, If we return an error here, then we should split `payFees` into two, `payMiner` and `paySharder`, so that they can fail independently of each other.